lib/ui/channel_buffers.dart

```diff
@@ -6,184 +6,334 @@
+    return channel.push(_StoredMessage(data, callback));
   }
 
-  bool _isEmpty(String channel) {
-    final _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    return queue == null || queue.isEmpty;
+  /// Sets the listener for the specified channel.
+  ///
+  /// When there is a listener, messages are sent immediately.
+  ///
+  /// Each channel may have up to one listener set at a time. Setting
+  /// a new listener on a channel with an existing listener clears the
+  /// previous one.
+  ///
+  /// Callbacks are invoked in their own stack frame and
+  /// use the zone that was current when the callback was
+  /// registered.
+  ///
+  /// ## Draining
+  ///
+  /// If any messages were queued before the listener is added,
+  /// they are drained asynchronously after this method returns.
+  ///
+  /// Each message is handled in its own microtask. No messages can
+  /// be queued by plugins while the queue is being drained, but any
+  /// microtasks queued by the handler itself will be processed before
+  /// the next message is handled.
+  ///
+  /// The draining stops if the listener is removed.
+  void setListener(String name, ChannelCallback callback) {
+    final _Channel channel = _channels.putIfAbsent(name, () => _Channel(kDefaultBufferSize));
+    channel.setListener(callback);
   }
 
-  /// Changes the capacity of the queue associated with the given channel.
+  /// Clears the listener for the specified channel.
   ///
-  /// This could result in the dropping of messages if newSize is less
-  /// than the current length of the queue.
-  void _resize(String channel, int newSize) {
-    _RingBuffer<_StoredMessage>? queue = _messages[channel];
-    if (queue == null) {
-      queue = _makeRingBuffer(newSize);
-      _messages[channel] = queue;
-    } else {
-      final int numberOfDroppedMessages = queue.resize(newSize);
-      if (numberOfDroppedMessages > 0) {
-        _Logger._printString('Dropping messages on channel "$channel" as a result of shrinking the buffer size.');
-      }
-    }
+  /// When there is no listener, messages on that channel are queued,
+  /// up to [kDefaultBufferSize] (or the size configured via the
+  /// control channel), and then discarded in a first-in-first-out
+  /// fashion.
+  void clearListener(String name) {
+    final _Channel? channel = _channels[name];
+    if (channel != null)
+      channel.clearListener();
   }
 
   /// Remove and process all stored messages for a given channel.
   ///
   /// This should be called once a channel is prepared to handle messages
   /// (i.e. when a message handler is setup in the framework).
-  Future<void> drain(String channel, DrainChannelCallback callback) async {
-    while (!_isEmpty(channel)) {
-      final _StoredMessage message = _pop(channel)!;
+  ///
+  /// The messages are processed by calling the given `callback`. Each message
+  /// is processed in its own microtask.
+  // TODO(ianh): deprecate once framework uses [setListener].
```

Comment:
If this method is going away we should explain the term "drain" in other dartdocs.

Reply:
The docs for setListener have a section on this.